### PR TITLE
Ft creatingsign up 158111211

### PR DESCRIPTION
**Pull request by alovega**
**What does this PR do?**
have CRUD methods for login sign out and signin

**Description of tasks completed**
have the following endpoints 
api.add_resource(UserRegister, '/auth/signup', endpoint='userregister')
api.add_resource(UserLogin, '/auth/login', endpoint='userlogin')
api.add_resource(UserLogoutRefresh,'/auth/logout/refresh',endpoint='userlogoutrefresh')
api.add_resource(UserLogoutAccess,'/auth/logout', endpoint='userlogout access')
api.add_resource(TokenRefresh,'/token/refresh')
**how should it be manually tested**
after cloning cd into it and RUN run.py in your cmd
*using postman test every endpoint above  with this header:
key: Content-Type value:application/json
to test authentication, add this header
Key:Authorization value:JWT token
**what are the relevant pivotal Tracker sories**
[https://www.pivotaltracker.com/story/show/158111211]

### CODE REVIEW

File: manage.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        commands = (
+            """
+            CREATE TABLE IF NOT EXISTS users(id SERIAL PRIMARY KEY, username VARCHAR NOT NULL UNIQUE, 
+            email VARCHAR(80) NOT NULL UNIQUE, password VARCHAR NOT NULL, is_admin VARCHAR(40) 
+            DEFAULT FALSE NOT NULL)""",
+            """
+            CREATE TABLE IF NOT EXISTS requests(id SERIAL PRIMARY KEY, title VARCHAR(50) NOT NULL,
+            description  VARCHAR(100) NOT NULL, category VARCHAR(40) NOT NULL,approve VARCHAR(40) 
+            DEFAULT FALSE, disapprove VARCHAR(40) DEFAULT FALSE,RESOLVE VARCHAR(40) DEFAULT FALSE, 
+            user_id INTEGER  REFERENCES users(id))
```
Comment (elc49) [inline manage.py:18]: This should be `user_id INTEGER, FOREIGN KEY (user_id) REFERENCES users (id)` @alovega

**Review by elc49 — COMMENTED** [review]
Check your code for creating relationships between users and requests table. I have commented on that line

File: maintenance/Request.py [hunk 2]
```diff
@@ -2,24 +2,24 @@
...
     def __str__(self):
         return self.request_title
 
 
 requests = []
-requests.append(RequestDao(request_id=1, request_title="laptop", request_description="laptop screen Repair",
-                           request_category="maintenance"))
-requests.append(RequestDao(request_id=2, request_title="window", request_description="window broken",
-                           request_category="maintenance"))
+maintenanceDao.insert_request(RequestDao(author = "kevin",title="laptop",description="laptop screen Repair",category="maintenance"))
```
Comment (daisymacharia) [inline maintenance/Request.py:21]: we don't need this anymore

File: models/models.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    def insert_user(self, UserDao):
+        sql = """INSERT INTO users(username,email,password) VALUES (%s,%s,%s)"""
+        # get connection
+        cur = self.connection.cursor()
+        # insert into database
+        cur.execute(sql, (UserDao.username, UserDao.email, UserDao.password,))
+        self.connection.commit()
+        cur.close()
+
+    def update_user(self, username,password,id):
```
Comment (daisymacharia) [inline models/models.py:27]: you could add update user role to admin

File: models/models.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        print(json.dumps(rows,indent=2))
+        return rows
+
+    def get_all(self):
+        cur = self.connection.cursor(cursor_factory=RealDictCursor)
+        cur.execute("SELECT id,username,email,password  from users")
+        rows = cur.fetchall ()
+        return rows
+
+    def check_user_exist(self, email):
```
Comment (daisymacharia) [inline models/models.py:75]: what is the difference between this method and `get_user_by_email` on line 53

File: models/models.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        self.connection.commit()
+        cur.close()
+
+    def delete_request(self, author):
+        cur = self.connection.cursor (cursor_factory=RealDictCursor)
+        cur.execute ("DELETE from users where author = %(author)s ",{'author':author})
+        rows_deleted = cur.rowcount
+        self.connection.commit()
+        print(json.dumps(rows_deleted,indent=2))
+        cur = self.connection.cursor (cursor_factory=RealDictCursor)
```
Comment (daisymacharia) [inline models/models.py:111]: is this instantiating of database connection necessary?

File: models/models.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        cur.close()
```
Comment (daisymacharia) [inline models/models.py:112]: this too

File: models/models.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        cur.execute ("SELECT id,username,email,password from users ")
+        rows = cur.fetchall ()
+        print (json.dumps (rows, indent=2))
+        return rows
+        print('rows')
+        cur.close()
+
+    def get_request_by_id(self, email):
+        cur = self.connection.cursor (cursor_factory=RealDictCursor)
+        cur.execute ("SELECT id,username,email,password from users where email = %(email)s ", {'email': email})
```
Comment (daisymacharia) [inline models/models.py:123]: this method queries the user table as opposed to the requests

File: models/models.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        rows = cur.fetchall ()
+        print(json.dumps(rows,indent=2))
+        return rows
+        cur.close()
+
+    def get_request_by_author(self, username,password):
+        cur = self.connection.cursor (cursor_factory=RealDictCursor)
+        cur.execute ("""SELECT id,username,email,password from users
```
Comment (daisymacharia) [inline models/models.py:131]: this query is similar to that of `get_user_by_username`

File: maintenance/__init__.py [hunk 9]
```diff
@@ -1,18 +1,31 @@
 from flask import Flask
 from flask_restful import Api
-
 from maintenance.Request import HelloWorld, RequestApi, RequestService
 from maintenance.User import UserRegister
 from maintenance.User import UserLogin
+from flask_jwt_extended import JWTManager
+from models import RevokedToken
```
Comment (daisymacharia) [inline maintenance/__init__.py:7]: you are importing `RevokedToken` from models but I do not see any method or object called `RevokedToken` in your models.

Comment (alovega) [reply maintenance/__init__.py:7]: ooh let me see

**Pull request merged**